File: frontend/src/mock/coupon.ts
/**
 * 美团神券 Mock 数据
 */

import Mock, { type MockRequestOptions } from 'mockjs';
import env from '../config/env';
import { successResponse } from './helpers';

if (env.USE_MOCK) {
  const mockCoupons = [
    {
      user_coupon_id: 1,
      coupon_id: 1,
      coupon_type: 'flash_sale',
      name: '闪购满29减5',
      description: '小象超市闪购专享',
      discount_type: 'fixed',
      discount_value: 5,
      min_order_amount: 29,
      applicable_scope: 'grocery',
      status: 'available',
      expires_at: new Date(Date.now() + 7 * 86400000).toISOString(),
    },
    {
      user_coupon_id: 2,
      coupon_id: 2,
      coupon_type: 'flash_sale',
      name: '闪购新客立减8',
      description: '闪购新客首单立减8元',
      discount_type: 'fixed',
      discount_value: 8,
      min_order_amount: 20,
      applicable_scope: 'grocery',
      status: 'available',
      expires_at: new Date(Date.now() + 10 * 86400000).toISOString(),
    },
    {
      user_coupon_id: 4,
      coupon_id: 4,
      coupon_type: 'delivery',
      name: '外卖满35减6',
      description: '美团外卖满35元减6元',
      discount_type: 'fixed',
      discount_value: 6,
      min_order_amount: 35,
      applicable_scope: 'delivery',
      status: 'available',
      expires_at: new Date(Date.now() + 5 * 86400000).toISOString(),
    },
    {
      user_coupon_id: 6,
      coupon_id: 6,
      coupon_type: 'cross_store',
      name: '跨店满50减10',
      description: '美团跨店满减',
      discount_type: 'fixed',
      discount_value: 10,
      min_order_amount: 50,
      applicable_scope: 'all',
      status: 'available',
      expires_at: new Date(Date.now() + 14 * 86400000).toISOString(),
    },
  ];

  Mock.mock(/\/api\/coupon\/user\/\d+(\?.*)?$/, 'get', (options: MockRequestOptions) => {
    const orderMatch = options.url?.match(/order_amount=([\d.]+)/);
    const orderAmount = orderMatch ? parseFloat(orderMatch[1]) : 0;

    const coupons = mockCoupons.map((c) => {
      const isApplicable = orderAmount >= c.min_order_amount;
      const estimated = isApplicable ? Math.min(c.discount_value, orderAmount) : 0;
      return { ...c, is_applicable: isApplicable, estimated_discount: estimated };
    });

    return successResponse({ coupons, total_count: coupons.length });
  });

  Mock.mock(/\/api\/coupon\/user\/\d+\/weekly-savings/, 'get', () => {
    return successResponse({
      weekly_saved_amount: 70.5,
      week_start: new Date().toISOString(),
      breakdown: {
        stock_used: 12.5,
        no_purchase: 28,
        coupon_saved: 8,
        near_expiry_used: 6.8,
        plan_saved: 15.2,
      },
      recent_logs: [
        {
          id: 1,
          event_type: 'plan_saved',
          saved_amount: 15.2,
          description: '选择省钱版补购方案，比标准版少花15.2元',
          created_at: new Date().toISOString(),
        },
      ],
    });
  });
}

File: backend/src/schemas/coupon.py
from pydantic import BaseModel, Field
from typing import List, Optional
from decimal import Decimal
from datetime import datetime


class CouponBaseItem(BaseModel):
    """神券模板"""
    id: int
    coupon_type: str = Field(..., description="flash_sale/delivery/cross_store")
    name: str
    description: Optional[str] = None
    discount_type: str = Field(..., description="fixed/percent")
    discount_value: Decimal
    min_order_amount: Decimal
    max_discount: Optional[Decimal] = None
    applicable_scope: str = Field(..., description="grocery/delivery/all")

    class Config:
        orm_mode = True


class UserCouponItem(BaseModel):
    """用户可用神券"""
    user_coupon_id: int
    coupon_id: int
    coupon_type: str
    name: str
    description: Optional[str] = None
    discount_type: str
    discount_value: Decimal
    min_order_amount: Decimal
    max_discount: Optional[Decimal] = None
    applicable_scope: str
    status: str
    expires_at: datetime
    estimated_discount: Optional[Decimal] = Field(None, description="预估可抵扣金额")
    is_applicable: bool = Field(False, description="当前订单是否可用")

    class Config:
        orm_mode = True


class MatchedCoupon(BaseModel):
    """方案匹配到的神券"""
    user_coupon_id: int
    coupon_id: int
    coupon_type: str
    name: str
    discount_amount: Decimal = Field(..., ge=0)
    min_order_amount: Decimal
    is_best: bool = False


class PriceComparisonBreakdown(BaseModel):
    """价格明细"""
    subtotal: Decimal
    coupon_discount: Decimal = Decimal("0")
    delivery_fee: Decimal = Decimal("0")
    final_price: Decimal


class PriceComparison(BaseModel):
    """自己买+做 vs 外卖+券后价 对比"""
    cook_self: PriceComparisonBreakdown
    takeout: PriceComparisonBreakdown
    recommended: str = Field(..., description="cook_self/takeout")
    savings_amount: Decimal = Field(..., description="推荐方案相对另一方案节省金额")
    savings_tip: str


class WeeklySavingsResponse(BaseModel):
    """本周库存管理省钱统计"""
    weekly_saved_amount: Decimal
    week_start: datetime
    breakdown: dict = Field(default_factory=dict, description="按事件类型汇总")
    recent_logs: List[dict] = Field(default_factory=list)


class UserCouponsResponse(BaseModel):
    """用户神券列表响应"""
    coupons: List[UserCouponItem]
    total_count: int
